src/main/scala/Table.scala

```diff
@@ -508,18 +507,70 @@ case class Table private (
   }
   warnings = warnings.concat(warningsFromColumns)
 
-  def validateRow(row: CSVRecord): WarningsAndErrors = {
+  def validateRow(row: CSVRecord): (
+      WarningsAndErrors,
+      List[Any],
+      List[(ForeignKeyWithTable, List[Any])],
+      List[(ForeignKeyWrapper, List[Any])]
+  ) = {
     var errors = Array[ErrorWithCsvContext]()
+    var primaryKeyValues = List[Any]()
+    var foreignKeyReferenceValues =
+      List[
+        (ForeignKeyWithTable, List[Any])
+      ]() // to store the validated referenced Table Columns values in each row
+    var foreignKeyValues =
+      List[
+        (ForeignKeyWrapper, List[Any])
+      ]() // to store the validated foreign key values in each row
     if (columns.nonEmpty) {
       for ((value, column) <- row.iterator.asScalaArray.zip(columns)) {
         //catch any exception here, possibly outOfBounds  and set warning too many values
-        val result = column.validate(value, row.getRecordNumber)
-        errors = errors.concat(result.errors)
-        warnings = warnings.concat(result.warnings)
+        val (errorsAndWarnings, v) = column.validate(
+          value,
+          row.getRecordNumber
+        )
+        val newValue: List[Any] = if (v.length == 1) {
+          List(v(0))
+        } else {
+          v.toList
+        } // newValue will be single element of type Any if separator is not specified. If separator is specified, newValue will be Array[Any]
+
+        errors = errors.concat(errorsAndWarnings.errors)
+        warnings = warnings.concat(errorsAndWarnings.warnings)
+
+        if (primaryKey.contains(column)) {
+          primaryKeyValues :+= newValue
+        }
+
+        for (foreignKeyReferenceObject <- foreignKeyReferences) {
+          if (
+            foreignKeyReferenceObject.referencedTableColumns.contains(column)
+          ) {
+            foreignKeyReferenceValues :+= (foreignKeyReferenceObject, newValue)
+          }
+        }
+
+        for (foreignKeyWrapperObject <- foreignKeys) {
+          if (foreignKeyWrapperObject.localColumns.contains(column)) {
+            foreignKeyValues :+= (foreignKeyWrapperObject, newValue)
+          }
+        }
       }
-      WarningsAndErrors(warnings, errors)
+
+      (
+        WarningsAndErrors(warnings, errors),
+        primaryKeyValues,
+        foreignKeyReferenceValues,
+        foreignKeyValues
+      )
     } else {
-      WarningsAndErrors(Array(), Array())
+      (
```

Comment:
Once you've replaced this with an object, we can just make it return Option.None here indead of empty stuff.

Reply:
Done